### PR TITLE
4-2-stable + ruby 2.4

### PR DESCRIPTION
**Pull request by matthewd**

### CODE REVIEW

**Comment by matthewd** [comment]
@pixeltrix do you have any thoughts on the smallest / safest / least back-branch-disruptive subset of https://github.com/rails/rails/commit/c9c5788a527b70d7f983e2b4b47e3afd863d9f48 & friends that would be sensible to backport?

I guess we're only behaving differently (either potentially slower, or potentially cachier, or both) if the configured `preserve_timezone` doesn't match the intrinsic behaviour of the running version, right? So `preserve_timezone=false` on pre-2.4 Rubies is zero cost?

**Comment by pixeltrix** [comment]
> I guess we're only behaving differently (either potentially slower, or potentially cachier, or both) if the configured preserve_timezone doesn't match the intrinsic behaviour of the running version, right? So preserve_timezone=false on pre-2.4 Rubies is zero cost?

Nearly zero - there's still the overhead of checking the config option at runtime. We could change it so that we rewrite the method at boot time, I guess.

As for backporting I think we should be able to do it fairly cleanly - IIRC there's a couple of gotchas but I can have a look at it this week.

File: Gemfile [hunk 1]
```diff
@@ -1,5 +1,7 @@
 source 'https://rubygems.org'
 
+ruby "~> #{RUBY_VERSION}" if ENV["TRAVIS"]
```
Comment (pixeltrix) [inline Gemfile]: Does rvm on Travis not blow up on this line? See rvm/rvm#3705

Comment (matthewd) [inline Gemfile]: I think that only matters if rvm is trying to parse the ruby version from the Gemfile.

If it's causing local problems, we can probably just obfuscate it a bit so rvm doesn't see it at all (e.g., make it a `send`).

Comment (pixeltrix) [inline Gemfile]: Is there an option to turn parsing of the Gemfile off?

**Comment by matthewd** [comment]
Merging ahead of https://github.com/rails/rails/pull/27553, which will get the build green again.

**Pull request merged**